Public/Src/Cache/ContentStore/Distributed/Sessions/DistributedContentSession.cs

```diff
@@ -83,10 +89,25 @@ private enum Counters
             UrgencyHint urgencyHint,
             Counter retryCounter)
         {
-            return PutCoreAsync(
-                operationContext,
-                (decoratedStreamSession, wrapStream) => decoratedStreamSession.PutFileAsync(operationContext, path, contentHash, realizationMode, operationContext.Token, urgencyHint, wrapStream),
-                session => session.PutFileAsync(operationContext, contentHash, path, realizationMode, operationContext.Token, urgencyHint));
+            return PerformPutFileGatedOperationAsync(operationContext, () => {
+                return PutCoreAsync(
+                    operationContext,
+                    (decoratedStreamSession, wrapStream) => decoratedStreamSession.PutFileAsync(operationContext, path, contentHash, realizationMode, operationContext.Token, urgencyHint, wrapStream),
+                    session => session.PutFileAsync(operationContext, contentHash, path, realizationMode, operationContext.Token, urgencyHint));
+            });
+        }
+
+        private Task<TResult> PerformPutFileGatedOperationAsync<TResult>(OperationContext operationContext, Func<Task<TResult>> func)
+        {
+            return _putFileGate.GatedOperationAsync((timeWaiting) =>
+            {
+                if (timeWaiting > Settings.PutFileWaitWarning)
+                {
+                    Tracer.Warning(operationContext, $"Spent {timeWaiting} waiting for PutFile gate, exceeding deadline of {Settings.PutFileWaitWarning}");

```

Comment:
I think it should be Info, not Warning, IMO.
I'm afraid of polluting the warning logs with all this...